src/native/shared/helpers.cc

```diff
@@ -1,15 +1,42 @@
+#include <cstdarg>
+#include <cstring>
+#include <android/set_abort_message.h>
+
 #include "helpers.hh"
+#include "java-interop-logger.h"
 
 using namespace xamarin::android;
 
 [[noreturn]] void
-Helpers::abort_application (bool log_location, std::source_location sloc) noexcept
+Helpers::abort_application (LogCategories category, const char *message, bool log_location, std::source_location sloc) noexcept
 {
+	// Log it, but also...
+	log_fatal (category, message);
+
+	// ...let android include it in the tombstone, debuggerd output, stack trace etc
+	android_set_abort_message (message);
```

Comment:
Do we not want `sloc` in this message either?

It feels like what we might want is C#-esque `${message}{(log_location ? $"\nabort at {sloc}" : "")}`, which would allow the location to be part of the abort message.

Question: can the message contain a newline? Does that work and what does the logcat output look like?

Reply:
Alas, it appears the maximum length of the message is 120 characters, so we can't really stuff the location info in there :(